arch/x86/zefi/printf.h

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <stdarg.h>
+#include <stdbool.h>
+
+/* Tiny, but not-as-primitive-as-it-looks implementation of something
+ * like s/n/printf().  Handles %d, %x, %p, %c and %s only, allows a
+ * "l" qualifier on %d and %x (and silently ignores one %s/%c/%p).
+ * Accepts, but ignores, field width and precision values that match:
+ * the regex: [0-9]*\.?[0-9]*
+ */
+
+struct _pfr {
+	char *buf;
+	int len;
+	int idx;
+};
+
+/* Set this function pointer to something that generates output */
+static void (*z_putchar)(int c);
+
+static void pc(struct _pfr *r, int c)
+{
+	if (r->buf) {
+		if (r->idx <= r->len)
+			r->buf[r->idx] = c;
+	} else {
+		z_putchar(c);
+	}
+	r->idx++;
+}
+
+static void prdec(struct _pfr *r, long v)
+{
+	if (v < 0) {
+		pc(r, '-');
+		v = -v;
+	}
+
+	char digs[11 * sizeof(long)/4];
+	int i = sizeof(digs) - 1;
+
+	digs[i--] = 0;
+	while (v || i == 9) {
+		digs[i--] = '0' + (v % 10);
+		v /= 10;
+	}
+
+	while (digs[++i])
+		pc(r, digs[i]);
+}
+
+static void endrec(struct _pfr *r)
+{
+	if (r->buf && r->idx < r->len)
+		r->buf[r->idx] = 0;
+}
+
+static int vpf(struct _pfr *r, const char *f, va_list ap)
```

Comment:
why not a bit more self-describing varialbe names?  like format or frmt instead of f etc...